src/OfflineTool.cpp

```diff
@@ -0,0 +1,503 @@
+*/
+//==============================================================================
+
+#include <OfflineTool.h>
+#include <RippleKey.h>
+#include <Serialize.h>
+#include <ripple/beast/core/SemanticVersion.h>
+#include <ripple/beast/unit_test.h>
+#include <beast/unit_test/dstream.hpp>
+#include <boost/algorithm/string/trim.hpp>
+#include <boost/filesystem.hpp>
+#include <boost/format.hpp>
+#include <boost/program_options.hpp>
+
+//------------------------------------------------------------------------------
+char const* const versionString =
+
+    //--------------------------------------------------------------------------
+    //  The build version number. You must edit this for each release
+    //  and follow the format described at http://semver.org/
+    //
+        "0.1.0"
+
+#if defined(DEBUG) || defined(SANITIZER)
+       "+"
+#ifdef DEBUG
+        "DEBUG"
+#ifdef SANITIZER
+        "."
+#endif
+#endif
+
+#ifdef SANITIZER
+        BEAST_PP_STR1_(SANITIZER)
+#endif
+#endif
+
+    //--------------------------------------------------------------------------
+    ;
+
+static
+int
+runUnitTests ()
+{
+    using namespace beast::unit_test;
+    beast::unit_test::dstream dout{std::cout};
+    reporter r{dout};
+    bool const anyFailed = r.run_each(global_suites());
+    if(anyFailed)
+        return EXIT_FAILURE;    //LCOV_EXCL_LINE
+    return EXIT_SUCCESS;
+}
+
+void
+doSerialize(std::string const& data)
+{
+    auto const tx = [&]
+    {
+        auto const json = offline::parseJson(data);
+        return json ? offline::makeObject(json) : boost::none;
+    }();
+    if (!tx)
+    {
+        std::cerr << "Unable to serialize \"" << data << "\"" <<
+            std::endl;
+        return;
+    }
+
+    auto const result = offline::serialize(*tx);
+
+    std::cout << result << std::endl;
+}
+
+void
+doDeserialize(std::string const& data)
+{
+    auto const fail = [&]
+    {
+        std::cerr << "Unable to deserialize \"" << data << "\"" <<
+            std::endl;
+    };
+    try
+    {
+        auto const result = offline::deserialize(boost::trim_copy(data));
+
+        if (result)
+            std::cout << result->getJson(0).toStyledString() << std::endl;
+        else
+            fail();
+    }
+    catch (std::exception const&)
+    {
+        fail();
+    }
+}
+
+void
+doSingleSign(std::string const& data,
+    boost::filesystem::path const& keyFile)
+{
+    using namespace offline;
+    auto const fail = [&]()
+    {
+        std::cerr << "Unable to sign \"" << data << "\"" << std::endl;
+    };
+    try
+    {
+        auto tx = make_sttx(boost::trim_copy(data));
+        if (tx)
+        {
+            auto const rippleKey = RippleKey::make_RippleKey(keyFile);
+
+            rippleKey.singleSign(*tx);
+
+            std::cout << tx->getJson(0).toStyledString() << std::endl;
+        }
+        else
+            fail();
+    }
+    catch (std::exception const&)
+    {
+        fail();
```

Comment:
In order to deal with the less-than-adequate information if there's a failure, you might consider reporting `std::exception::what()` in addition to the current text.  You could also report at what point the attempt failed (serializing, creating the rippleKey, or signing) by using finer granularity of the `try/catch` or by using a state variable that tracks which step is being handled inside the `try`.